### PR TITLE
tests: deflake TestTaskRunner_StatsHook_Periodic

### PR DESCRIPTION
**Pull request by notnoop**
This PR deflakes TestTaskRunner_StatsHook_Periodic tests and adds backoff when the driver closes the channel.

TestTaskRunner_StatsHook_Periodic is currently the most flaky test - failing ~4% of the time (20 out of 486 workflows). A sample failure: https://app.circleci.com/pipelines/github/hashicorp/nomad/14028/workflows/957b674f-cbcc-4228-96d9-1094fdee5b9c/jobs/128563 .

This change has two components:

First, it updates the StatsHook so that it backs off when stats channel is closed. In the context of the test where the mock driver emits a single stats update and closes the channel, the test may make tens of thousands update during the period.  In real context, if a driver doesn't implement the stats handler properly or when a task finishes, we may generate way too many Stats queries in a tight loop.  Here, the backoff reduces these queries.  I've added a failing test that shows 154,458 stats updates within 500ms in https://app.circleci.com/pipelines/github/hashicorp/nomad/14092/workflows/50672445-392d-4661-b19e-e3561ed32746/jobs/129423 .

Second, the test ignores the first stats update after a task exit.  Due to the asynchronicity of updates and channel/context use, it's possible that an update is enqueued while the test marks the task as exited, resulting into a spurious update.

### CODE REVIEW

File: client/allocrunner/taskrunner/stats_hook.go [hunk 1]
```diff
@@ -99,7 +99,15 @@ MAIN:
...
-				goto MAIN
+				// backoff if driver closes channel, potentially
+				// because task shutdown or because driver
+				// doesn't implement channel interval checking
+				select {
+				case <-time.After(h.interval):
+					goto MAIN
+				case <-ctx.Done():
+					return
+				}
```
Comment (notnoop) [inline client/allocrunner/taskrunner/stats_hook.go:110]: I've decided to unilaterily back off with interval regardless of the reason.  We could be a bit more discerning and wait for a fraction of interval, or only what's left of the interval, but I didn't see the additional complexity worth it, as this is a workaround a driver misbehavior case.

File: client/allocrunner/taskrunner/stats_hook_test.go [hunk 2]
```diff
@@ -150,10 +159,19 @@ func TestTaskRunner_StatsHook_Periodic(t *testing.T) {
 	require.NoError(h.Exited(context.Background(), nil, nil))
 
 	// Should *not* get another update in ~500ms (see interval above)
+	// we may get a single update due to race with exit
+	timeout := time.After(2 * interval)
+	firstUpdate := false
```
Comment (drewbailey) [inline client/allocrunner/taskrunner/stats_hook_test.go:164]: should this be `firstUpdate := true`?

**Review by drewbailey — APPROVED** [review]
LGTM! :tada:

File: client/allocrunner/taskrunner/stats_hook_test.go [hunk 3]
```diff
@@ -150,10 +159,19 @@ func TestTaskRunner_StatsHook_Periodic(t *testing.T) {
...
+	timeout := time.After(2 * interval)
+	firstUpdate := true
+
+WAITING:
 	select {
 	case ru := <-su.Ch:
+		if firstUpdate {
+			firstUpdate = false
+			goto WAITING
+		}
```
Comment (notnoop) [inline client/allocrunner/taskrunner/stats_hook_test.go:172]: Here, we also can be a bit more discerning to ensure that no stats update happens after the task is marked as exited, but It's not fully clear to me how we can completely rule out the race between exiting and channel sending; so I decided to ignore the first update after the exit as spurious update.

**Review by shoenig — APPROVED** [review]

**Pull request merged**